tests/commands/test_get.py

```diff
@@ -0,0 +1,432 @@
+    ['type=laptop', 'type=headphones'],
+    ['num=16', 'num=16'],
+    ['num=8', 'num=16'],
+    ['num=8.*', 'num=16.*'],
+    ['num']])
+def test_get_filter_errors(repo: Repo, filters: list[str]) -> None:
+    """
+    Test that `onyo get --filter KEY=VALUE --depth 999` retrieves the expected
+    files.
+
+    Parametrized filters contain Tuple([filters], n_assets expected to match).
+    """
+    cmd = ['onyo', 'get', '--filter', *filters, '--depth', '999', '-H']
+    ret = subprocess.run(cmd, capture_output=True, text=True)
+
+    assert ret.stderr
+    assert not ret.stdout
+    assert ret.returncode == 1
+
+
+@pytest.mark.repo_contents(*convert_contents([
+    ('laptop_apple_macbookpro.1', {'num': 8, 'str': 'foo', 'bool': True}),
+    ('one/laptop_dell_precision.2', {'num': '16', 'str': 'bar', 'bool': False}),
+    ('one/two/headphones_apple_pro.3', {'num': '8', 'str': 'bar', 'bool': 'True'})]))
+@pytest.mark.parametrize('keys', [
+    ['type', 'make', 'model', 'serial'],
+    ['unset', 'type', 'unset2', 'make'],
+    ['num', 'str', 'bool'],
+    ['TyPe', 'MAKE', 'moDEL', 'NuM', 'STR'],
+    []])
+def test_get_keys(repo: Repo, keys: list) -> None:
+    """
+    Test that `onyo get --depth 3 --keys x y z` retrieves the expected keys.
+    """
+    raw_assets = [
+        ('laptop_apple_macbookpro.1', {'num': 8, 'str': 'foo', 'bool': True}),
+        ('one/laptop_dell_precision.2', {'num': '16', 'str': 'bar', 'bool': False}),
+        ('one/two/headphones_apple_pro.3', {'num': '8', 'str': 'bar', 'bool': 'True'})]
+
+    cmd = ['onyo', 'get', '--depth', '999', '--keys', *keys, '-H']
+    ret = subprocess.run(cmd, capture_output=True, text=True)
+    output = [output.split('\t') for output in ret.stdout.split('\n')][:-1]
+
+    # default keys returned if no keys were specified
+    if not keys:
+        keys = repo.default_keys
+
+    # Get all the key values and make sure they match
+    for line in output:
+        asset = raw_assets[[a[0] for a in raw_assets].index(line[-1])][1]
+
+        # add type, make, model, serial from asset name
+        asset = asset | dict(zip(
+            ['type', 'make', 'model', 'serial'],
+            re.split('[_.]', Path(line[-1]).name)))
+
+        for i, key in enumerate(keys):
+            # convert raw asset values to str because output type is str
+            assert str(asset.get(key, '<unset>')) == line[i]
+
+    assert len(output) == len(repo.assets)
+    assert not ret.stderr
+    assert ret.returncode == 0
+
+
+@pytest.mark.repo_contents(*convert_contents([
+    ('laptop_apple_macbookpro.1', {}),
+    ('one/laptop_dell_precision.2', {}),
+    ('one/two/headphones_apple_pro.3', {}),
+    ('one/two/three/headphones_apple_pro.4', {}),
+    ('one/two/three/four/headphones_apple_pro.5', {})]))
+@pytest.mark.parametrize('depth_expected', [
+    ('0', 1), ('1', 2), ('2', 3), ('3', 4), ('4', 5), ('999', 5), ('-1', 0)])
+def test_get_depth(repo: Repo, depth_expected: tuple[str, int]) -> None:
```

Comment:
`depth` defaults to 0, thus only checking the cwd for assets. I set `depth = 999` so that all assets are included, including the ones not in root. This ensures that the asset name is derived properly from the path.

The test that addresses depth tests for various different values, and I just added a `test_get_default` to test that the defaults work as intended.

Reply:
My whole point is (here and for other flags below) that the tests should *not* assume any specific defaults to be set in `main.py`, but exclusively test for what the user might do, and then test that the behavior is like we wish it to be.

So if we have flags, and it is possible to not enter them, then it should be tested that the user gets what was requested. If we have flags that can conflict, we should have a test that checks that the conflict leads to a loud error.

Reply:
I can adjust this and then remove any directory depth from the parametrized assets. `test_get_all()` tests the use of multiple flags combined, so rather than setting `depth = 999`, I will set it as optional for those tests.